packages/api-plugin-orders/src/util/orderValidators/getFinalFulfillmentGroups.js

```diff
@@ -0,0 +1,148 @@
+import Random from "@reactioncommerce/random";
+import buildOrderItem from "../buildOrderItem.js";
+import addInvoiceToGroup from "../addInvoiceToGroup.js";
+import addShipmentMethodToGroup from "../addShipmentMethodToGroup.js";
+import addTaxesToGroup from "../addTaxesToGroup.js";
+import compareExpectedAndActualTotals from "../compareExpectedAndActualTotals.js";
+import getSurchargesForGroup from "../getSurchargesForGroup.js";
+
+/**
+ * @summary Create fulfillment groups for a potential order
+ * @param {Object} context - The application context
+ * @param {Object} inputData - The input data
+ * @param {String} orderId ID of existing or new order to which this group will belong
+ * @param {String} [accountId] ID of account that is placing or already did place the order
+ * @param {Object} [additionalItems] Additional items if any
+ * @param {Object} [billingAddress] The primary billing address for the order, if known
+ * @param {String} [cartId] ID of the cart from which the order is being placed, if applicable
+ * @param {String} currencyCode Currency code for all money values
+ * @param {Number} [discountTotal] Calculated discount total
+ * @param {Object} [fulfillmentGroups] Fulfillment group input
+ * @param {Object} cart Cart object
+ * @returns {Promise<Object>} Object with surcharge, total, shippingAddress and finalFulfillment groups
+ */
+export default async function getFinalFulfillmentGroups(context, inputData) {
+  const {
+    orderId,
+    accountId,
+    additionalItems,
+    billingAddress,
+    cartId,
+    currencyCode,
+    discountTotal,
+    fulfillmentGroups,
+    cart
+  } = inputData;
+
+  // Create array for surcharges to apply to order, if applicable
+  // Array is populated inside `fulfillmentGroups.map()`
+  const orderSurcharges = [];
+  let shippingAddressForPayments = null;
+  let orderTotal = 0;
+
+  // Add more props to each fulfillment group, and validate/build the items in each group
+
+  const finalFulfillmentGroups = await Promise.all(fulfillmentGroups.map(async (inputGroup) => {
+    const { data, items, selectedFulfillmentMethodId, shopId, totalPrice: expectedGroupTotal, type } = inputGroup;
+    const group = {
+      _id: Random.id(),
+      address: data ? data.shippingAddress : null,
+      shopId,
+      type,
+      workflow: { status: "new", workflow: ["new"] }
+    };
+
+    // Build the final order item objects. As part of this, we look up the variant in the system and make sure that
+    // the price is what the caller expects it to be.
+    if (items) {
+      group.items = await Promise.all(items.map((inputItem) => buildOrderItem(context, { currencyCode, inputItem, cart })));
+    } else {
+      group.items = [];
+    }
```

Comment:
This may work and look a bit better?
```suggestion
    group.items = await Promise.all((items || []).map((inputItem) => buildOrderItem(context, { currencyCode, inputItem, cart })));
```